accepted/better-obsoletion/better-obsoletion.md

```diff
@@ -0,0 +1,345 @@
+  than having to turn off all obsoletion diagnostics or having to resort to
+  suppressing all call sites individually.
+
+* **Documentation pages are easily accessible**. Developers must be able to go
+  from the diagnostic in the error list straight to the documentation that
+  provides the details.
+
+### Non-Goals
+
+* **Detecting specific usage patterns**. Sometimes, an API isn't obsoleted, just
+  a particular combination of arguments or specific values. If we need to
+  identify these, we'll author a custom analyzer.
+
+* **Being able to obsolete APIs in already shipped versions of .NET**. The
+  attribute must be applied to the API surface that the compiler sees and thus
+  must be present in the reference assemblies. We can only obsolete in new
+  versions of the platform. In other words, this doesn't support a sidecar file
+  that can be applied to an existing version of .NET.
+
+* **Being able to remove APIs in later versions**. See
+  [the discussion](#why-is-being-able-to-delete-apis-not-a-goal) in the Q&A
+  section.
+
+## Design
+
+The design proposal piggybacks on the existing `ObsoleteAttribute` that is
+already supported by all .NET compilers. The idea is to expand this a little to
+support the desired scenarios. Alternatives are listed afterwards.
+
+### API Changes
+
+We'd add the following properties to `ObsoleteAttribute`:
+
+```C#
+namespace System
+{
+    public sealed class ObsoleteAttribute : Attribute
+    {
+        // Existing:
+        //
+        // public ObsoleteAttribute();
+        // public ObsoleteAttribute(string message);
+        // public ObsoleteAttribute(string message, bool error);
+        // public bool IsError { get; }
+        // public string Message { get; }
+
+        // New:
+        public string DiagnosticId { get; set; }
+        public string UrlFormat { get; set; }
+        public string Url => UrlFormat == null ? null : string.Format(UrlFormat, DiagnosticId);
+    }
+}
+```
+
+Instead of taking the URL directly, the API takes a format string. This allows
+having a generic URL that includes the diagnostic ID. This avoids having to
+repeat the ID twice and thus making a copy & paste mistake.
+
+### Compiler Changes
+
+The compilers would be enlightened about these new properties and use them as
+follows:
+
+* If `DiagnosticId` is `null`, use the existing diagnostic ID (e.g. `CS0618` in
+  C#). Otherwise, use the specified warning ID.
+* If `Url` is not `null`, use it as the diagnostic link when rendering them in
+  the IDE.
+* The compiler should assume that `Url` and `DiagnosticId` are independent
+  features, in other words both can be used, either, and neither.
+
+### Samples
+
+Here is how the attribute would be applied:
+
+```C#
+internal static class Obsolete
+{
+    public const string Url = "http://aka.ms/obsolete/{0}";
+
+    // NOTE: Keep this sorted by the the ID's value,
+    //       not the name of the field!
+    public const string SecureStringId = "BCL0001";
```

Comment:
I was surprised to see all BCL code being exposed (A) in one place, and (B) on the Obsolete type itself. I think a better model would be to have no codes exposed on this type whatsoever. Your goal is to establishing a naming scheme, which is good, but this design is missing an opportunity to communicate that. Instead, we could adopt the pattern of exposing types called [component]"Obsolete", like "BclObjsolete", "SystemObsolete" or "AspNetCoreObsolete".  We'd create such a type now with whatever name we liked.

I'm hoping we can come up with a better name than "BCL" since it doesn't show up anywhere in the platform. I'd prefer "SystemObsolete" but that is incredibly close to "System.Obsolete". Another option is "CoreFXObsolete", tying it to repo name. That may be a bad idea for other reasons, but it would be easy to communicate and for developers to grok / implement.

Reply:
> on the Obsolete type itself.

Note that this is a different internal type, *not* `ObsoleteAttribute`.

The idea is to store the IDs and URL in a central location on a per-library basis to prevent typos and allow changing the URL from one place. (Basically this class addresses [my concerns on the earlier PR](https://github.com/dotnet/designs/pull/60#issuecomment-479606367).)

Reply:
> I was surprised to see all BCL code being exposed (A) in one place

I only did this to showcase internal plumbing to help us centralize the diagnostic IDs. This wouldn't be exposed to the consumer.